### PR TITLE
Data directories configurable separately per service

### PR DESCRIPTION
**Pull request by lhoss**
I introduced new variables which allow for full flexibility in configuring the (local) data/storage dirs, and at the same time re-using the same base_*_dir variables in the new defaults

Motivation: fulfill following requirements:
* configure *a set* of hdfs-datanodes directories and different from the default base_data_dir (p.eg `/data`)
* configure a kafka-log directory, different from the default base_data_dir (p.eg `/data`)
* configure ranger audit spool dirs, in different folder than the base_log_dir (p.eg `/var/log`)
* configure different nifi dirs (commit coming)

Changes inspired from: https://github.com/hortonworks/ansible-hortonworks/pull/75 (that PR I will close now)

### CODE REVIEW

**Comment by agriffaut** [comment]
Hi @lhoss, thanks for your PR !
I'll review it as soon as possible..

**Comment by lhoss** [comment]
great @agriffaut . Btw, until tomorrow I've done >1 test deployment.. and might add small new changes (incl. a dozen nifi changes).

A btw: Question I wanted to ask:
Would you mind us/me moving all the new vars (including your base_*_dir vars) into the blueprint role `defaults/main.yml` (vars only used in blueprint anyway, so we won't overcrowd the playbook/all vars) ?!

**Comment by agriffaut** [comment]
Yes moving to blueprint role defaults, is definitively a good idea !

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 1]
```diff
@@ -464,18 +467,16 @@
...
         "dfs.namenode.inode.attributes.provider.class" : "org.apache.ranger.authorization.hadoop.RangerHdfsAuthorizer",
 {% endif %}
 {% if rangerkms_hosts|length > 0 %}
         "dfs.encryption.key.provider.uri" : "kms://http@{% for kmshost in rangerkms_hosts %}{{ hostvars[kmshost]['ansible_fqdn'] }}{% if not loop.last %};{% endif %}{% endfor %}:9292/kms",
 {% endif %}
-        "dfs.datanode.data.dir" : "{{ base_data_dir }}/hdfs/data",
-        "dfs.datanode.failed.volumes.tolerated" : "0",
-        "dfs.replication" : "3"
+        "dfs.datanode.data.dir" : "{{ hdfs_dfs_datanode_data_dir }}",
+        "dfs.datanode.failed.volumes.tolerated" : "{{ hdfs_dfs_datanode_failed_volumes_tolerated | default('0') }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 2]
```diff
@@ -464,18 +467,16 @@
...
+        "dfs.replication" : "{{ hdfs_dfs_replication | default('3') }}"
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 3]
```diff
@@ -919,27 +920,32 @@
...
       }
     },
     {
       "kafka-broker": {
 {% if rangeradmin_hosts|length > 0 and ranger_options.enable_plugins|default(no) %}
         "authorizer.class.name" : "org.apache.ranger.authorization.kafka.authorizer.RangerKafkaAuthorizer",
 {% endif %}
-        "zookeeper.session.timeout.ms" : "30000",
-        "log.dirs" : "{{ base_data_dir }}/kafka-logs"
+        "default.replication.factor": "{{ kafka_broker_default_replication_factor | default('1') }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 4]
```diff
@@ -919,27 +920,32 @@
...
+        "log.dirs" : "{{ kafka_broker_log_dirs }}",
+        "log.retention.bytes": "{{ kafka_broker_log_retention_bytes | default('-1') }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 5]
```diff
@@ -919,27 +920,32 @@
...
+        "log.retention.hours": "{{ kafka_broker_log_retention_hours | default('168') }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 6]
```diff
@@ -919,27 +920,32 @@
...
+        "num.partitions": "{{ kafka_broker_num_partitions | default('1') }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 7]
```diff
@@ -919,27 +920,32 @@
...
+        "offsets.topic.replication.factor": "{{ kafka_broker_offsets_topic_replication_factor | default('3') }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 8]
```diff
@@ -919,27 +920,32 @@
...
+        "zookeeper.session.timeout.ms" : "{{ kafka_broker_zookeeper_session_timeout_ms | default('30000') }}"
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: This is an undefined var.. and is out of the scope of the feature

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 [hunk 9]
```diff
@@ -445,12 +445,15 @@
...
         "namenode_opt_maxnewsize" : "384m",
         "namenode_opt_newsize" : "384m",
-        "hdfs_log_dir_prefix" : "{{ base_log_dir }}/hadoop",
+        "hdfs_log_dir_prefix" : "{{ hdfs_log_dir_prefix }}",
         "namenode_backup_dir" : "{{ base_tmp_dir }}/upgrades"
       }
     },
     {
       "hdfs-site" : {
+        "dfs.namenode.checkpoint.dir" : "{{ hdfs_dfs_namenode_checkpoint_dir }}",
```
Comment (agriffaut) [inline playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: Nice catch !

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 (hunk 1 above)
Comment (lhoss) [reply playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: ok I undo all those changes here (though thx to the `| default(...)` the vars don't need to be defined at all)
and add it sep. PR

File: playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2 (hunk 9 above)
Comment (lhoss) [reply playbooks/roles/ambari-blueprint/templates/blueprint_dynamic.j2]: yeah, hope we did not miss another one 👍

**Comment by lhoss** [comment]
> Some properties are out of the scope of this feature..

I undid all of them (for this PR)
And added a final commit for tuneable NiFI/Registry directories (tested with a HDP-3.1,HDF-3.3.1 AWS deployment) 👍

**Comment by agriffaut** [comment]
Good job ! :clap:

**Comment by lfrancke** [comment]
This is fabulous and just what we'd need.

Any plans to merge it back upstream?

**Comment by agriffaut** [comment]
Hi @lfrancke, this is indeed a PR on hortonworks repository.
Please see https://github.com/hortonworks/ansible-hortonworks/pull/72

**Comment by lfrancke** [comment]
Excellent, thank you!

I saw that PR but thought it was an "old" one. Sorry, I missed the connection.

Either way: Good news, thanks for your work.

**Comment by lhoss** [comment]
> Any plans to merge it back upstream?

Hi @lfrancke , don't hesitate to ask that question in the upstream PR 👍  Would be great to get this (
( and other nice PRs by @agriffaut  and myself there) merged at some point.
ps: When I've more time in 2weeks I plan to open some discussion issues .. I've some idea for larger refactorings (esp. wondering to rethink the set_variables logic, and to have all vars overrideable by group_vars)

**Pull request merged**